README.md
# H2N5 [![Build Status](https://github.com/aschampion/h2n5/actions/workflows/ci.yml/badge.svg)](https://github.com/aschampion/h2n5/actions/workflows/ci.yml/)

H2N5 is:
- ~~An Australian strain of avian flu~~
- ~~[(E)-Hydrazinylidenehydrazinylidene]azanide~~
- HTTP 2 N5: A simple program to serve N5 datasets over HTTP as tiled image stacks

## Minimum supported Rust version (MSRV)

Stable 1.56

## Quick start

Installation options:
- Compile and install using cargo: `cargo install h2n5`
- Download a precompiled binary from [the latest GitHub releases](https://github.com/aschampion/h2n5/releases/latest)
- Install a precompiled binary using [cargo-binstall](https://github.com/ryankurte/cargo-binstall): `cargo binstall h2n5`

```
h2n5 path/to/my.n5
curl http://127.0.0.1:8088/tile/group/dataset/0_1/512_256/1/2/3.jpg?q=80
```

Tile URLs are constructed as:

```
http://[bind_address]:[port]/tile/[n5_dataset]/[slicing_dims]/[tile_size]/[coordinates].[format]?[query parameters]
```

For example, the URL:

```
http://127.0.0.1:8088/tile/group/dataset/0_1/512_256/1/2/3.jpg?q=80
```

Will slice a 512 px by 256 px tile from the `group/dataset` dataset along axes `0` (as tile X) and `1` (as tile Y), respectively. The returned tile will start at voxel coordinates [1, 2, 3], be encoded as JPEG, with a quality of 80.

For more options, see the command line help:

```
h2n5 -h
```

## Notes

- PNG (`png`) and JPEG (`jpg`|`jpeg`) encoding formats are supported.
- Currently, only grayscale tiles are returned (by slicing remaining dimensions as singletons) or grayscale + alpha or RGBA by packing these scalar values across channels using the `pack` query parameter. Slicing a third dimension for RGB(A) channels (e.g., `slicing_dims` as `0_1_4`) will be supported, and is currently parsed correctly, but not yet implemented for tile encoding.

### Supported data type, channel packing, and encoding formats

|        | gray (default) | graya      | rgba       |
|--------|----------------|------------|------------|
| UINT8  | JPEG PNG       |            |            |
| UINT16 | PNG            | JPEG PNG   |            |
| UINT32 |                | PNG        | JPEG PNG   |
| UINT64 |                |            | PNG        |
| FLOAT32|                | PNG        | JPEG PNG   |
| FLOAT64|                |            | PNG        |

### Logging

Logs are output to stderr using standard Rust `log` and `env_logger` in the `actix_web` scope. By default only major errors at the `WARN` level or above are shown. For example, to show all requests:

```
RUST_LOG=info h2n5
```

For more information, see the [`env_logger`](https://docs.rs/env_logger) and [`actix_web` logging middleware](https://actix.rs/docs/errors/) documentation.

## Development

Version bumping (including changelog release section versioning) is handled by [cargo-release](https://github.com/crate-ci/cargo-release), e.g.:

```prompt
cargo release --dev-version minor
```

## License

Licensed under either of

- Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
- MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted for inclusion in the work by you, as defined in the Apache-2.0 license, shall be dual licensed as above, without any additional terms or conditions.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).


<!-- next-header -->
## [Unreleased]

## [0.1.9] - 2022-06-13
### Changed
- Updated from actix-web 1 to 4, in preparation for more asynchronous IO.

## [0.1.8] - 2022-06-07
### Changed
- Updates to dependencies for bug fixes.

## [0.1.7] - 2020-10-22
### Fixed
- Fixed a soundness error in a dependency that caused a panic on Rust 1.48 and
  greater (see
  [this issue](https://github.com/contain-rs/linked-hash-map/pull/100)).
- Reduced allocation frequency and size for encoded tiles.

## [0.1.6] - 2020-07-18
### Fixed
- Fixed a panic when a block's dataset-level bounds overlapped a tile request
  but the block's actual bounds did not.

## [0.1.5] - 2020-07-13
### Added
- An optional LRU memory cache for tiles can be enabled with the `--tile-cache`
  option. The size of the cache (in tiles) globally can be configured with the
  `--tile-cache-size` option. HTTP server caching should be preferred for
  production deployments, but this option is useful for temporary cases.

### Fixed
- Fixes from dependencies include a bug that caused n5 to corrupt the version
  attribute when opening datasets and extremely rare cases where PNG tiles
  would be corrupted during compression.

## [0.1.4] - 2019-11-15
### Added
- An optional LRU memory cache for blocks can be enabled with the
  `--ds-block-cache` option. The size of the cache (in blocks) per-dataset can
  be configured with the `--ds-block-cache-size` option.
- Scalar value bytes can be packed into channels by adding a `pack` query
  parameter to request URLs. For example: `?pack=rgba`.
- `FLOAT32` and `FLOAT64` datasets are now supported via channel packing.
- The maximum size (in bytes) that will be preallocated for a tile before
  decoding can be configured with `--max-tile-prealloc`.
- Logging is now enabled via `env_logger`.

### Changed
- Actix-web and many other dependencies underwent major update that include
  security issues.
- Cumulative optimizations can improve request response times by 40%.

### Fixed
- A bug that could cause a panic on some platforms and environments has been
  fixed.

<!-- next-url -->
[Unreleased]: https://github.com/aschampion/h2n5/compare/v0.1.9...HEAD
[0.1.9]: https://github.com/aschampion/h2n5/compare/v0.1.8...v0.1.9
[0.1.8]: https://github.com/aschampion/h2n5/compare/0.1.7...v0.1.8
[0.1.7]: https://github.com/aschampion/h2n5/compare/0.1.6...0.1.7
[0.1.6]: https://github.com/aschampion/h2n5/compare/0.1.5...0.1.6
[0.1.5]: https://github.com/aschampion/h2n5/compare/0.1.4...0.1.5
[0.1.4]: https://github.com/aschampion/h2n5/compare/0.1.3...0.1.4


src/cache.rs
use std::collections::HashMap;
use std::io::Error;
use std::sync::RwLock;

use anymap::{
    any::Any,
    Map,
};
use log::debug;
use lru_cache::LruCache;
use n5::prelude::*;
use n5::{
    ReadableDataBlock,
    ReinitDataBlock,
};

type DatasetBlockCache<BT> = LruCache<GridCoord, Option<VecDataBlock<BT>>>;

struct BlockCache<BT: ReflectedType> {
    blocks: RwLock<HashMap<String, RwLock<DatasetBlockCache<BT>>>>,
}

impl<BT: ReflectedType> BlockCache<BT> {
    fn new() -> Self {
        BlockCache {
            blocks: RwLock::new(HashMap::new()),
        }
    }

    fn clear(&mut self) {
        self.blocks.write().unwrap().clear();
    }
}

pub struct N5CacheReader<N: N5Reader> {
    reader: N,
    blocks_capacity: usize,
    attr_cache: RwLock<HashMap<String, DatasetAttributes>>,
    cache: Map<dyn Any + Send + Sync>,
}

impl<N: N5Reader> N5CacheReader<N> {
    pub fn wrap(reader: N, blocks_capacity: usize) -> Self {
        let mut cache = Map::new();
        cache.insert(BlockCache::<i8>::new());
        cache.insert(BlockCache::<i16>::new());
        cache.insert(BlockCache::<i32>::new());
        cache.insert(BlockCache::<i64>::new());
        cache.insert(BlockCache::<u8>::new());
        cache.insert(BlockCache::<u16>::new());
        cache.insert(BlockCache::<u32>::new());
        cache.insert(BlockCache::<u64>::new());
        cache.insert(BlockCache::<f32>::new());
        cache.insert(BlockCache::<f64>::new());
        Self {
            reader,
            blocks_capacity,
            attr_cache: RwLock::new(HashMap::new()),
            cache,
        }
    }

    #[allow(dead_code)]
    pub fn clear(&mut self) {
        self.attr_cache.write().unwrap().clear();
        self.cache.get_mut::<BlockCache<i8>>().unwrap().clear();
        self.cache.get_mut::<BlockCache<i16>>().unwrap().clear();
        self.cache.get_mut::<BlockCache<i32>>().unwrap().clear();
        self.cache.get_mut::<BlockCache<i64>>().unwrap().clear();
        self.cache.get_mut::<BlockCache<u8>>().unwrap().clear();
        self.cache.get_mut::<BlockCache<u16>>().unwrap().clear();
        self.cache.get_mut::<BlockCache<u32>>().unwrap().clear();
        self.cache.get_mut::<BlockCache<u64>>().unwrap().clear();
        self.cache.get_mut::<BlockCache<f32>>().unwrap().clear();
        self.cache.get_mut::<BlockCache<f64>>().unwrap().clear();
    }
}

impl<N: N5Reader> N5Reader for N5CacheReader<N> {
    fn get_version(&self) -> Result<n5::Version, Error> {
        self.reader.get_version()
    }

    fn get_dataset_attributes(&self, path_name: &str) -> Result<n5::DatasetAttributes, Error> {
        {
            if let Some(data_attrs) = self.attr_cache.read().unwrap().get(path_name) {
                return Ok(data_attrs.clone());
            }
        }

        let data_attrs = self.reader.get_dataset_attributes(path_name)?;
        self.attr_cache
            .write()
            .unwrap()
            .insert(path_name.to_owned(), data_attrs.clone());

        Ok(data_attrs)
    }

    fn exists(&self, path_name: &str) -> Result<bool, Error> {
        self.reader.exists(path_name)
    }

    fn get_block_uri(&self, path_name: &str, grid_position: &[u64]) -> Result<String, Error> {
        self.reader.get_block_uri(path_name, grid_position)
    }

    fn read_block<T>(
        &self,
        path_name: &str,
        data_attrs: &DatasetAttributes,
        grid_position: GridCoord,
    ) -> Result<Option<VecDataBlock<T>>, Error>
    where
        VecDataBlock<T>: DataBlock<T> + ReadableDataBlock,
        T: ReflectedType,
    {
        let cache = self.cache.get::<BlockCache<T>>().unwrap();

        if cache.blocks.read().unwrap().get(path_name).is_none() {
            cache
                .blocks
                .write()
                .unwrap()
                .entry(path_name.to_owned())
                .or_insert_with(|| RwLock::new(LruCache::new(self.blocks_capacity)));
        }

        {
            // Done in a separate scope to not hold any locks while reading
            // blocks on a cache miss.
            // Have to explicitly write these out to satisfy pre-NLL rust borrowck.
            let cache_lock = cache.blocks.read().unwrap();
            let ds_cache = &cache_lock[path_name];
            let mut ds_block_cache = ds_cache.write().unwrap();

            if let Some(block) = ds_block_cache.get_mut(&grid_position[..]) {
                debug!(
                    "Block cache hit: {path_name} {grid_position:?}",
                    path_name = path_name,
                    grid_position = grid_position
                );
                return Ok(block.clone());
            }
        }

        let block = self
            .reader
            .read_block(path_name, data_attrs, grid_position.clone())?;
        let ds_cache = &cache.blocks.read().unwrap()[path_name];
        ds_cache
            .write()
            .unwrap()
            .insert(grid_position, block.clone());

        Ok(block)
    }

    fn read_block_into<
        T: ReflectedType,
        B: DataBlock<T> + ReinitDataBlock<T> + ReadableDataBlock,
    >(
        &self,
        path_name: &str,
        data_attrs: &DatasetAttributes,
        grid_position: GridCoord,
        block: &mut B,
    ) -> Result<Option<()>, Error> {
        let cache = self.cache.get::<BlockCache<T>>().unwrap();

        if cache.blocks.read().unwrap().get(path_name).is_none() {
            cache
                .blocks
                .write()
                .unwrap()
                .entry(path_name.to_owned())
                .or_insert_with(|| RwLock::new(LruCache::new(self.blocks_capacity)));
        }

        {
            // Done in a separate scope to not hold any locks while reading
            // blocks on a cache miss.
            // Have to explicitly write these out to satisfy pre-NLL rust borrowck.
            let cache_lock = cache.blocks.read().unwrap();
            let ds_cache = &cache_lock[path_name];
            let mut ds_block_cache = ds_cache.write().unwrap();

            if let Some(maybe_block) = ds_block_cache.get_mut(&grid_position[..]) {
                return match maybe_block {
                    Some(existing_block) => {
                        block.reinitialize_with(existing_block);
                        Ok(Some(()))
                    }
                    None => Ok(None),
                };
            }
        }

        let maybe =
            self.reader
                .read_block_into(path_name, data_attrs, grid_position.clone(), block)?;
        let ds_cache = &cache.blocks.read().unwrap()[path_name];
        let maybe_block = maybe.map(|_| {
            VecDataBlock::new(
                block.get_size().into(),
                block.get_grid_position().into(),
                block.get_data().into(),
            )
        });
        ds_cache.write().unwrap().insert(grid_position, maybe_block);

        Ok(maybe)
    }

    fn block_metadata(
        &self,
        path_name: &str,
        data_attrs: &DatasetAttributes,
        grid_position: &[u64],
    ) -> Result<Option<DataBlockMetadata>, Error> {
        self.reader
            .block_metadata(path_name, data_attrs, grid_position)
    }

    fn list_attributes(&self, path_name: &str) -> Result<serde_json::Value, Error> {
        self.reader.list_attributes(path_name)
    }
}

impl<N: N5Lister> N5Lister for N5CacheReader<N> {
    fn list(&self, path_name: &str) -> Result<Vec<String>, Error> {
        self.reader.list(path_name)
    }
}


src/tiling.rs
use std::num::ParseIntError;
use std::str::FromStr;

use n5::ndarray::prelude::*;
use n5::smallvec::smallvec;
use n5::smallvec::SmallVec;
use n5::{
    DatasetAttributes,
    N5Reader,
    ReadableDataBlock,
    ReflectedType,
    ReinitDataBlock,
};

#[derive(Debug, Eq, PartialEq, Hash)]
pub struct SlicingDims {
    pub plane_dims: [u32; 2],
    pub channel_dim: Option<u32>,
}

#[derive(Copy, Clone, Debug, Eq, PartialEq, Hash)]
pub struct TileSize {
    pub w: u32,
    pub h: u32,
}

impl FromStr for TileSize {
    type Err = ParseIntError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let coords: SmallVec<[&str; 3]> = s.split(',').collect();

        let w = coords[0].parse::<u32>()?;
        let h = coords[1].parse::<u32>()?;

        Ok(TileSize { w, h })
    }
}

impl std::fmt::Display for TileSize {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{},{}", self.w, self.h)
    }
}

#[derive(Debug, Eq, PartialEq, Hash)]
pub struct TileSpec {
    pub n5_dataset: String,
    pub slicing_dims: SlicingDims,
    pub tile_size: TileSize,
    pub coordinates: SmallVec<[u64; 6]>,
}

pub fn read_tile<T, N: N5Reader>(
    n: &N,
    data_attrs: &DatasetAttributes,
    spec: &TileSpec,
) -> Result<ndarray::Array<T, ndarray::Dim<ndarray::IxDynImpl>>, std::io::Error>
where
    n5::VecDataBlock<T>: n5::DataBlock<T> + ReinitDataBlock<T> + ReadableDataBlock,
    T: ReflectedType + num_traits::identities::Zero,
{
    // Express the spec tile as an N-dim bounding box.
    let mut size = smallvec![1u64; data_attrs.get_dimensions().len()];
    size[spec.slicing_dims.plane_dims[0] as usize] = u64::from(spec.tile_size.w);
    size[spec.slicing_dims.plane_dims[1] as usize] = u64::from(spec.tile_size.h);
    if let Some(dim) = spec.slicing_dims.channel_dim {
        size[dim as usize] = data_attrs.get_dimensions()[dim as usize];
    }
    let bbox = BoundingBox::new(spec.coordinates.clone(), size);

    // Read the N-dim slab of blocks containing the tile from N5.
    n.read_ndarray::<T>(&spec.n5_dataset, data_attrs, &bbox)
}


src/encoding.rs
use std::io::Write;
use std::str::FromStr;

use image::ImageEncoder;
use n5::{
    DatasetAttributes,
    N5Reader,
    ReadableDataBlock,
    ReflectedType,
    ReinitDataBlock,
};

use crate::tiling::read_tile;
use crate::TileImageSpec;

#[derive(Debug, Eq, PartialEq, Hash)]
pub struct JpegParameters {
    pub(crate) quality: u8,
}

impl Default for JpegParameters {
    fn default() -> JpegParameters {
        JpegParameters { quality: 100 }
    }
}

#[derive(Debug, Eq, PartialEq, Hash)]
pub enum EncodingFormat {
    Jpeg(JpegParameters),
    Png,
}

impl EncodingFormat {
    fn encode<W: Write>(
        &self,
        writer: &mut W,
        bytes: &[u8],
        width: u32,
        height: u32,
        image_color_type: image::ColorType,
    ) -> Result<(), image::ImageError> {
        match *self {
            EncodingFormat::Jpeg(ref p) => {
                let mut encoder =
                    image::codecs::jpeg::JpegEncoder::new_with_quality(writer, p.quality);
                encoder.encode(bytes, width, height, image_color_type)
            }
            EncodingFormat::Png => {
                let encoder = image::codecs::png::PngEncoder::new(writer);
                encoder.write_image(bytes, width, height, image_color_type)
            }
        }
    }

    pub fn content_type(&self) -> &'static str {
        match *self {
            EncodingFormat::Jpeg(_) => "image/jpeg",
            EncodingFormat::Png => "image/png",
        }
    }
}

impl FromStr for EncodingFormat {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "jpg" | "jpeg" => Ok(EncodingFormat::Jpeg(JpegParameters::default())),
            "png" => Ok(EncodingFormat::Png),
            _ => Err(()),
        }
    }
}

#[derive(Debug, Eq, PartialEq, Hash)]
pub enum ChannelPacking {
    Gray,
    GrayA,
    Rgba,
}

impl ChannelPacking {
    fn num_channels(&self) -> u8 {
        match self {
            ChannelPacking::Gray => 1,
            ChannelPacking::GrayA => 2,
            ChannelPacking::Rgba => 4,
        }
    }
}

impl Default for ChannelPacking {
    fn default() -> Self {
        ChannelPacking::Gray
    }
}

impl FromStr for ChannelPacking {
    type Err = (); // TODO

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s.to_lowercase().as_str() {
            "gray" => Ok(ChannelPacking::Gray),
            "graya" => Ok(ChannelPacking::GrayA),
            "rgba" => Ok(ChannelPacking::Rgba),
            _ => Err(()),
        }
    }
}

#[derive(thiserror::Error, Debug)]
pub enum EncodingError {
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    Image(#[from] image::ImageError),
}

pub fn read_and_encode<T, N: N5Reader, W: Write>(
    n: &N,
    data_attrs: &DatasetAttributes,
    spec: &TileImageSpec,
    writer: &mut W,
) -> Result<(), EncodingError>
where
    n5::VecDataBlock<T>: n5::DataBlock<T> + ReinitDataBlock<T> + ReadableDataBlock,
    T: ReflectedType + num_traits::identities::Zero,
{
    let tile = read_tile(n, data_attrs, &spec.tile)?;
    encode_tile(tile, spec, writer)
}

// TODO: Single channel only.
pub fn encode_tile<T, W: Write>(
    tile: ndarray::Array<T, ndarray::Dim<ndarray::IxDynImpl>>,
    spec: &TileImageSpec,
    writer: &mut W,
) -> Result<(), EncodingError>
where
    T: ReflectedType,
{
    let image_color_type = match spec.tile.slicing_dims.channel_dim {
        Some(_dim) => {
            // TODO: match RGB/RGBA based on dimensions of dim.
            // Permute slab so that channels dimension is at end.
            unimplemented!()
        }
        None => {
            let bits_per_channel = 8 / spec.packing.num_channels() * std::mem::size_of::<T>() as u8;
            match bits_per_channel {
                // Wow, this is so much better than just specifying my channels and BPC! /s
                8 => match spec.packing {
                    ChannelPacking::Gray => image::ColorType::L8,
                    ChannelPacking::GrayA => image::ColorType::La8,
                    ChannelPacking::Rgba => image::ColorType::Rgba8,
                },
                16 => match spec.packing {
                    ChannelPacking::Gray => image::ColorType::L16,
                    ChannelPacking::GrayA => image::ColorType::La16,
                    ChannelPacking::Rgba => image::ColorType::Rgba16,
                },
                _ => {
                    return Err(std::io::Error::new(
                        std::io::ErrorKind::InvalidData,
                        "Packed bits per channel must be 8 or 16",
                    )
                    .into())
                }
            }
        }
    };

    let data = if spec.tile.slicing_dims.plane_dims[0] > spec.tile.slicing_dims.plane_dims[1] {
        // Note, this works correctly because the slab is f-order.
        tile.into_iter().cloned().collect()
    } else {
        tile.into_raw_vec()
    };

    // Get the image data as a byte slice.
    let bytes: &[u8] = unsafe { as_u8_slice(&data) };

    spec.format
        .encode(
            writer,
            bytes,
            spec.tile.tile_size.w,
            spec.tile.tile_size.h,
            image_color_type,
        )
        .map_err(Into::into)
}

// Get the byte slice of a vec slice in a wrapper function
// so that the lifetime is bound to the original slice's lifetime.
unsafe fn as_u8_slice<T>(s: &[T]) -> &[u8] {
    std::slice::from_raw_parts(
        s as *const [T] as *const u8,
        s.len() * std::mem::size_of::<T>(),
    )
}


src/main.rs
use std::cell::RefCell;
use std::collections::HashMap;
use std::path::PathBuf;
use std::str::FromStr;
use std::sync::{
    Arc,
    RwLock,
};

use actix_cors::Cors;
use actix_web::middleware::Condition;
use actix_web::web::Bytes;
use actix_web::web::Data;
use actix_web::web::Query;
use actix_web::*;
use log::debug;
use lru_cache::LruCache;
use n5::filesystem::N5Filesystem;
use n5::smallvec::SmallVec;
use n5::{
    DataType,
    N5Reader,
};
use structopt::StructOpt;

mod cache;
mod encoding;
mod tiling;

use cache::N5CacheReader;
use encoding::*;
use tiling::*;

/// Trait for types that can be configured by URL query string parameters.
trait QueryConfigurable {
    fn configure(&mut self, params: &HashMap<String, String>);
}

impl QueryConfigurable for JpegParameters {
    fn configure(&mut self, params: &HashMap<String, String>) {
        if let Some(q) = params.get("q").and_then(|q| q.parse::<u8>().ok()) {
            self.quality = q;
        }
    }
}

impl QueryConfigurable for EncodingFormat {
    #[allow(unknown_lints)]
    fn configure(&mut self, params: &HashMap<String, String>) {
        if let EncodingFormat::Jpeg(ref mut p) = self {
            p.configure(params);
        }
    }
}

impl ChannelPacking {
    fn from_query(params: &HashMap<String, String>) -> Result<Self, ()> {
        if let Some(pack) = params.get("pack") {
            Self::from_str(pack)
        } else {
            Ok(ChannelPacking::default())
        }
    }
}

#[derive(thiserror::Error, Debug)]
pub enum TileImageSpecError {
    #[error("Invalid value for tiling parameter: {0}")]
    InvalidValue(#[from] std::num::ParseIntError),
    #[error("Tiling request path was malformed")]
    MalformedPath,
    #[error("Unknown encoding format")]
    UnknownEncodingFormat,
    #[error("Unknown channel packing")]
    UnknownChannelPacking,
}

impl ResponseError for TileImageSpecError {
    fn error_response(&self) -> HttpResponse {
        HttpResponse::build(http::StatusCode::BAD_REQUEST).body(self.to_string())
    }
}

/// Specifies a tile to slice from a dataset and an image format to encode it.
#[derive(Debug, PartialEq, Eq, Hash)]
pub struct TileImageSpec {
    tile: TileSpec,
    format: EncodingFormat,
    packing: ChannelPacking,
}

impl FromStr for TileImageSpec {
    type Err = TileImageSpecError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        lazy_static::lazy_static! {
                static ref RE: regex::Regex = regex::Regex::new(concat!(
                r"^(?P<dataset>.*)/(?P<slicing>\d+_\d+(_\d+)?)/",
                r"(?P<tile_size>\d+_\d+)(?P<coords>(/\d+)+)\.(?P<format>.+)$"
            ))
            .expect("Impossible: regex is valid");
        }

        let caps = RE.captures(s).ok_or(TileImageSpecError::MalformedPath)?;

        let n5_dataset = caps["dataset"].into();
        let mut sd_vals = caps["slicing"].split('_').map(u32::from_str);

        let slicing_dims = SlicingDims {
            plane_dims: [sd_vals.next().unwrap()?, sd_vals.next().unwrap()?],
            channel_dim: sd_vals.next().transpose()?,
        };

        let mut ts_vals = caps["tile_size"].split('_').map(u32::from_str);

        let tile_size = TileSize {
            w: ts_vals.next().unwrap()?,
            h: ts_vals.next().unwrap()?,
        };

        let coordinates = caps["coords"]
            .split('/')
            .filter(|n| !str::is_empty(*n))
            .map(u64::from_str)
            .collect::<Result<SmallVec<_>, _>>()?;

        let format = EncodingFormat::from_str(&caps["format"])
            .map_err(|_| TileImageSpecError::UnknownEncodingFormat)?;

        let packing = ChannelPacking::default();

        Ok(TileImageSpec {
            tile: TileSpec {
                n5_dataset,
                slicing_dims,
                tile_size,
                coordinates,
            },
            format,
            packing,
        })
    }
}

#[allow(unknown_lints)]
async fn tile<N: N5Reader>(
    state: Data<AppState<N>>,
    req: HttpRequest,
    query: Query<HashMap<String, String>>,
) -> Result<HttpResponse> {
    let spec = {
        let mut spec = TileImageSpec::from_str(&req.match_info()["spec"])?;
        spec.format.configure(&query);
        spec.packing = ChannelPacking::from_query(&query)
            .map_err(|_| TileImageSpecError::UnknownChannelPacking)?;
        spec
    };

    debug!("{spec:?}", spec = spec);

    if spec.tile.tile_size.w > state.max_tile_size.w
        || spec.tile.tile_size.h > state.max_tile_size.h
    {
        return Ok(HttpResponse::BadRequest()
            .reason("Maximum tile size exceeded")
            .finish());
    }

    if let Some(cache) = &state.tile_cache {
        let mut cache_lock = cache.write().unwrap();
        if let Some(tile) = cache_lock.get_mut(&spec) {
            debug!("Tile cache hit");
            return Ok(HttpResponse::Ok()
                .content_type(spec.format.content_type())
                .body(tile.clone()));
        }
    }

    let n = &state.n5;
    let data_attrs = n.get_dataset_attributes(&spec.tile.n5_dataset)?;
    let mut buffer = state.tile_buffer.borrow_mut();
    buffer.clear();

    use std::ops::DerefMut;
    match *data_attrs.get_data_type() {
        DataType::UINT8 => {
            read_and_encode::<u8, _, _>(&**n, &data_attrs, &spec, buffer.deref_mut())?
        }
        DataType::UINT16 => {
            read_and_encode::<u16, _, _>(&**n, &data_attrs, &spec, buffer.deref_mut())?
        }
        DataType::UINT32 => {
            read_and_encode::<u32, _, _>(&**n, &data_attrs, &spec, buffer.deref_mut())?
        }
        DataType::UINT64 => {
            read_and_encode::<u64, _, _>(&**n, &data_attrs, &spec, buffer.deref_mut())?
        }
        DataType::FLOAT32 => {
            read_and_encode::<f32, _, _>(&**n, &data_attrs, &spec, buffer.deref_mut())?
        }
        DataType::FLOAT64 => {
            read_and_encode::<f64, _, _>(&**n, &data_attrs, &spec, buffer.deref_mut())?
        }
        _ => {
            return Ok(HttpResponse::NotImplemented()
                .reason("Data type does not have an image renderer implemented")
                .finish())
        }
    }

    // Use `as_slice().to_vec()` rather than `.clone()` to allocate the smallest
    // `Vec` possible.
    let tile: Bytes = buffer.as_slice().to_vec().into();
    let content_type = spec.format.content_type();
    if let Some(cache) = &state.tile_cache {
        cache.write().unwrap().insert(spec, tile.clone());
    }

    Ok(HttpResponse::Ok().content_type(content_type).body(tile))
}

impl ResponseError for EncodingError {
    fn error_response(&self) -> HttpResponse {
        match *self {
            Self::Io(ref e) => e.error_response(),
            Self::Image(_) => {
                HttpResponse::build(http::StatusCode::BAD_REQUEST).body(self.to_string())
            }
        }
    }
}

/// Serve N5 datasets over HTTP as tiled image stacks.
///
/// Configuration options.
#[derive(StructOpt, Debug, Clone)]
#[structopt(name = "h2n5")]
struct Options {
    /// Bind address
    #[structopt(short = "b", long = "bind-address", default_value = "127.0.0.1")]
    bind_address: std::net::IpAddr,

    /// Bind port
    #[structopt(short = "p", long = "port", default_value = "8088")]
    port: u16,

    /// Number of worker threads for handling requests.
    /// By default, the number of CPU cores is used.
    #[structopt(short = "t", long = "threads")]
    threads: Option<usize>,

    /// Allow wildcard cross-origin requests (CORS).
    ///
    /// This does not yet configure specific allowed origins,
    /// only a wildcard accept. This is most useful for
    /// development purposes. For specific CORS policies, proxy
    /// behind another HTTP server.
    #[structopt(short = "c", long = "cors")]
    cors: bool,

    /// N5 root path
    #[structopt(name = "N5_ROOT_PATH", parse(from_os_str), default_value = ".")]
    root_path: PathBuf,

    /// Maximum tile buffer preallocation
    #[structopt(long = "max-tile-prealloc", default_value = "1000000")]
    max_tile_prealloc: usize,

    /// Maximum tile size
    #[structopt(long = "max-tile-size", default_value = "4096,4096")]
    max_tile_size: TileSize,

    /// Whether to cache blocks
    #[structopt(long = "ds-block-cache")]
    ds_block_cache: bool,

    /// Cache size (in blocks) per dataset
    #[structopt(long = "ds-block-cache-size", default_value = "1024")]
    ds_block_cache_size: usize,

    /// Whether to cache tile images
    #[structopt(long = "tile-cache")]
    tile_cache: bool,

    /// Cache size (in tiles) globally
    #[structopt(long = "tile-cache-size", default_value = "1024")]
    tile_cache_size: usize,
}

type TileCache = LruCache<TileImageSpec, Bytes>;

struct AppState<N: N5Reader> {
    n5: Arc<N>,
    // `RefCell` is fine because this state is per-thread.
    tile_buffer: RefCell<Vec<u8>>,
    max_tile_size: TileSize,
    tile_cache: Option<Arc<RwLock<TileCache>>>,
}

// Must manually implement Clone because derive on generics requires them
// to be Clone even when inside an Arc (open bug in Rust 26925).
impl<N: N5Reader> Clone for AppState<N> {
    fn clone(&self) -> Self {
        AppState {
            n5: self.n5.clone(),
            tile_buffer: self.tile_buffer.clone(),
            max_tile_size: self.max_tile_size,
            tile_cache: self.tile_cache.clone(),
        }
    }
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    let opt = Options::from_args();
    let root_path = opt.root_path.to_str().ok_or_else(|| {
        std::io::Error::new(std::io::ErrorKind::InvalidData, "Paths must be UTF-8")
    })?;
    let n5 = N5Filesystem::open(root_path)?;
    let max_tile_prealloc = opt.max_tile_prealloc;
    let max_tile_size = opt.max_tile_size;
    let tile_cache = if opt.tile_cache {
        Some(Arc::new(RwLock::new(TileCache::new(opt.tile_cache_size))))
    } else {
        None
    };

    if opt.ds_block_cache {
        let state = AppState {
            n5: Arc::new(N5CacheReader::wrap(n5, opt.ds_block_cache_size)),
            tile_buffer: RefCell::new(Vec::with_capacity(max_tile_prealloc)),
            max_tile_size,
            tile_cache,
        };
        run_server(opt, state).await
    } else {
        let state = AppState {
            n5: Arc::new(n5),
            tile_buffer: RefCell::new(Vec::with_capacity(max_tile_prealloc)),
            max_tile_size,
            tile_cache,
        };
        run_server(opt, state).await
    }
}

async fn run_server<N: N5Reader + Send + Sync + 'static>(
    opt: Options,
    state: AppState<N>,
) -> std::io::Result<()> {
    let cors = opt.cors;
    env_logger::init();

    let mut server = HttpServer::new(move || {
        App::new()
            .app_data(web::Data::new(state.clone()))
            .wrap(actix_web::middleware::Logger::default())
            .service(web::resource("/tile/{spec:.*}").route(web::get().to(tile::<N>)))
            .wrap(Condition::new(cors, Cors::permissive()))
    })
    .bind((opt.bind_address, opt.port))?;
    if let Some(threads) = opt.threads {
        server = server.workers(threads);
    }
    server.run().await
}

#[cfg(test)]
mod tests {
    use super::{
        EncodingFormat,
        JpegParameters,
        SlicingDims,
        TileImageSpec,
        TileSize,
    };
    use n5::smallvec::SmallVec;
    use std::str::FromStr;

    #[test]
    fn test_tile_spec_parsing() {
        let ts = TileImageSpec::from_str("my_test/dataset/0_1/512_512/3/2/1.jpg").unwrap();

        assert_eq!(ts.tile.n5_dataset, "my_test/dataset");
        assert_eq!(
            ts.tile.slicing_dims,
            SlicingDims {
                plane_dims: [0u32, 1],
                channel_dim: None,
            }
        );
        assert_eq!(ts.tile.tile_size, TileSize { w: 512, h: 512 });
        assert_eq!(ts.tile.coordinates, SmallVec::from([3u64, 2, 1]));
        assert_eq!(ts.format, EncodingFormat::Jpeg(JpegParameters::default()));
    }
}
